### DISCUSSION TITLE
Battery Management System/Solar charge controller

### DISCUSSION THREAD

**Discussion by pkkrusty** [arendst/Tasmota#16257] [Show and tell]
*Disclaimer: Lithium batteries can be dangerous if you don't know what you're doing

A friend who is re-using electric car batteries to make large lithium installations had a bunch of Mercedes BMS modules based on Analog Devices LTC6802-1. He loaned me a few to play with, and I've cobbled together a simple solar battery charger based on cheap AliExpress modules. The system can handle around 15amps and works with systems from 4s to 10s

The LTC6802-1 communicates via SPI, and Tasmota scripting can do basic SPI communication. 
![IMG_4140](https://user-images.githubusercontent.com/79770016/185479451-96c68691-95dd-4123-9f8a-0d751e8b1916.jpg)
Solar input and load output are controlled with a combination of MOSFET switch and mechanical relay in parallel. The MOSFET switches on first, then mechanical relay, then MOSFET switches off. This minimizes arcing and allows higher-than-rated voltages. (30v on the little 10a relays can go up to 100v if arc suppression is achieved) [Article](https://sound-au.com/articles/hybrid-relays.htm#s2)

Solar panel is a 350w 42v Voc, 38v Vmp, which works well for systems up to 8s. In my case, 9s because I limit charging to 4.0v instead of 4.2v. MOSFET for solar input is FRN120 (100v max), and D4184 for output (30v max). No MPPT or even PWM is happening for solar input. Either on or off.

I used a ESP8266 Wemos Mini D1 clone (4mb) compiled with 160mhz option. Pin assignments are as follows:
<img width="365" alt="Screen Shot 2022-08-18 at 20 11 04" src="https://user-images.githubusercontent.com/79770016/185485690-0e6ea857-757c-4688-8cac-5b511a84cbe1.png">
I used default hardware SPI since it sounds like software SPI can't handle 1mhz clock speed. I used a INA226 with an external 15a shunt, calibrated using a DC current clamp multimeter, to measure bus voltage and current going into and out of the battery. INA226 can handle voltage up to 36v if measured on the high side, so the system would be limited to 8s using that module. A PZEM-017 can handle voltages of up to 300v, and would be a good choice for higher voltage systems.

User_config_override.h has the following defines for scripting, SPI, I2C. Heap hovers around 20 with these options.
```
#define USE_SCRIPT                               // Add support for script (+17k code)
 #define USE_SCRIPT_FATFS 4                     // Script: Add FAT FileSystem Support
 #define SUPPORT_MQTT_EVENT                     // Support trigger event with MQTT subscriptions (+3k5 code)
//#define USE_SML_M                                  // Not used
#define USE_SCRIPT_WEB_DISPLAY
#define USE_SCRIPT_JSON_EXPORT
//#define USE_SML_SCRIPT_CMD              // Not used
//#define USE_SCRIPT_SERIAL                   // Not used
#define USE_SCRIPT_TIMER
//#define SCRIPT_LARGE_VNBUFF            // Not used, but may be necessary as the script grows
#define USE_SCRIPT_SPI
#define USE_SCRIPT_SUB_COMMAND
#define USE_UFILESYS
#define UFSYS_SIZE 4096

#ifdef MAXFILT
#undef MAXFILT
#endif
#define MAXFILT 8

#define USE_I2C
#define I2CDRIVERS_32_63       0xFFFFFFFF          // Enable I2CDriver32 to I2CDriver63
  #define USE_INA226                             // [I2cDriver35] Enable INA226 (I2C address 0x40, 0x41 0x44 or 0x45) Low voltage and current sensor (+2k3 code)

#define USE_SPI                                  // Hardware SPI using GPIO12(MISO), GPIO13(MOSI) and GPIO14(CLK) in addition to two user selectable GPIOs(CS and DC)
```
Script currently takes about 70% of the defined buffer. In theory, can use 6.5k buffer if more is needed.
```
>D
SB=4096
IP=192.168.1.46
m:cvr=0 19
m:cfr=0 7
m:wcfr=0 7
m:cv=0 12
m:cmnd=0 1
dcc=0
dcchold=1
cvhold=0
cntr=0
min=4
max=0
minc=0
maxc=0
numcells=6
mci=0
;write config trigger
wcfg=0
avg=0
;communications counter
cmcntr=0
;Voltage under value 125 * 16 * 1.5mV = 3.000v
vuv=125
;Voltage over value 167 * 16 * 1.5mV = 4.008v
vov=167
maxdiscnct=4
maxrecnct=3.95
mindiscnct=3.2
minrecnct=3.3
res=0
vstr=""
rlcll1=0
rlcll2=0
rlcll3=0
blncstrt=3.9

>B
cvr[0]=0
cfr[0]=0
cfr[1]=2
wcfr[0]=0
;write config register with command (1) and first byte (00000101)
wcfr[1]=1
wcfr[2]=5
cmnd[0]=0
;initialize SPI interface with MOSI, MISO, CLK
res=spi(0 14 13 12)
;initialize SPI interface with CS (GPIO1) and speed (1mhz)
res=spi(1 1 1)
;reduce backlog intercommand delay to 50ms
=>setoption34 50
rlcll2=0
;create cell mask depending on numcells
mci=4095-(1<<(numcells+2)-1)

>S
;Read cell values command and clear cvr array to ensure things are written correctly
cvr[1]=4
for cntr 2 19 1
cvr[cntr]=0
next
;Start A/D conversion command
cmnd[1]=0x10
; Read config registers command and clear cfr array
cfr[1]=2
for cntr 2 7 1
cfr[cntr]=0
next
;Clear write config array
for cntr 1 7 1
wcfr[cntr]=0
next

;if discharge tribble changed since last loop, re-write config register to activate correct discharge. Initial write of config register is necessary to activate the LTC6802-1 chip. Re-write once every 60 seconds in case LTC6802-1 communication is lost (it will go back to default state)
if ((dcchold!=dcc and upsecs%30==0) or upsecs%60==0) {
wcfg=1
}

wcfr[1]=1
wcfr[2]=5
wcfr[3]=(dcc&255)
wcfr[4]=((dcc>>8)|(mci<<4&240))
wcfr[5]=(mci>>4)
wcfr[6]=vuv
wcfr[7]=vov

;print %wcfr[1]%,%wcfr[2]%,%wcfr[3]%,%wcfr[4]%,%wcfr[5]%,%wcfr[6]%,%wcfr[7]%
if (wcfg==1) {
;spi command to send completed register
res=spi(2 1 wcfr 7 1)
;print sent %wcfr[1]%,%wcfr[2]%,%wcfr[3]%,%wcfr[4]%,%wcfr[5]%,%wcfr[6]%,%wcfr[7]%
wcfg=0
}
;Command to start A/D conversion (takes 21ms)
res=spi(2 1 cmnd 1 1)
delay(22)
;Command to read cell values
res=spi(2 1 cvr 19 1)
;read cell voltage registers
delay(2)
;Command to read config register
res=spi(2 1 cfr 7 1)

;Counter to prove >w section is still functioning and provide visual uptime
rlcll2+=1
if rlcll2>100000 {
rlcll2=0
}

;average cell voltage values
avg=0
for cntr 1 numcells 1
avg+=cv[cntr]
next
avg/=numcells

vstr=""

;cv[1]=((cvr[6]%16)<<8)|(cvr[5]&255)*.0015
;rlcll1=((cvr[3]%16)<<8)|(cvr[2]&255)*.0015
;rlcll2=(cvr[4]*16)|((cvr[3]>>4)&15)*.0015
;cv[2]=(cvr[7]*16)|((cvr[6]>>4)&15)*.0015
;cv[3]=((cvr[9]%16)<<8)|(cvr[8]&255)*.0015
;cv[4]=(cvr[10]*16)|((cvr[9]>>4)&15)*.0015
;cv[5]=((cvr[12]%16)<<8)|(cvr[11]&255)*.0015
;cv[6]=(cvr[13]*16)|((cvr[12]>>4)&15)*.0015
;Programmatically calc 12-bit cell voltages from 8-bit registers
for cntr 1 (numcells-1) 2
cv[cntr]=((cvr[cntr*1.5+4.5]%16)<<8)|(cvr[(cntr*1.5+3.5)]&255)*.0015
next
for cntr 2 (numcells) 2
cv[cntr]=(cvr[(cntr*1.5+4)]*16)|((cvr[(cntr*1.5+3)]>>4)&15)*.0015
next
;Adjust cell 1 due to bad BMS measurement
rlcll3=cv[1]
;Quadratic regression analysis shows r^2=.9998, so pretty good fit, linear would also work
rlcll1=5.947*cv[1]-(0.7506*cv[1]*cv[1])-7.576
cv[1]=rlcll1

;Calc max and min cells
for cntr 1 numcells 1
if cntr==1 {
min=cv[cntr]
max=cv[cntr]
}
if cv[cntr]>=max {
max=cv[cntr]
maxc=cntr
}
if cv[cntr]<=min {
min=cv[cntr]
minc=cntr
}
next

dcchold=dcc

dcc=0
;Calc discharge tribble for first cell separately since c1,c2,c3 are all connected to gnd
cvhold=(cv[1]-.02)
if (cvhold>avg and cv[1]>blncstrt) {
vstr+="d"
dcc=dcc|(7)
}
else {
if cv[1]<=avg {
dcc=(dcc&4088)
}
vstr+="r"
}
;Finish discharge tribble programmatically
for cntr 2 numcells 1
cvhold=cv[cntr]-0.02
if (cvhold>avg and cv[cntr]>blncstrt) {
vstr+="d"
dcc=dcc|(1<<(cntr+1))
}
else { 
if cv[cntr]<=avg {
dcc=dcc&(4095-(1<<(cntr+1)))
}
vstr+="r"
}
next

if upsecs%60==0 {
;print DCC=%dcc%, DCCa=%cfr[3]%, avg=%avg%, vstr=%vstr%, wcfr1=%wcfr[2]%, wcfr2=%wcfr[3]%, wcfr3=%wcfr[4]%, wcfr4=%wcfr[5]%
;print cfr1=%cfr[2]%, cfr2=%cfr[3]%, cfr3=%cfr[4]%, cfr4=%cfr[5]%
;print cvr=%cvr[2]%,%cvr[3]%,%cvr[4]%,%cvr[5]%,%cvr[6]%,%cvr[7]%,%cvr[8]%,%cvr[9]%,%cvr[10]%,%cvr[11]%,%cvr[12]%,%cvr[13]%,%cvr[14]%,%cvr[15]%,%cvr[16]%,%cvr[17]%,%cvr[18]%,%cvr[19]%

;Check voltages to change input/output relays (SSR switches loads on and off, and mechanical relay takes over after bouncing and arcing risk is over)
if ((max>maxdiscnct) and (pwr[3]==1)) {
=>backlog power1 1; power3 0; power1 0
}
if ((max<maxrecnct) and (pwr[3]==0)) {
=>backlog power1 1; power3 1; power1 0
}
if ((min<mindiscnct or (time>=240 and time<=1080)) and (pwr[4]==1)) {
=>backlog power2 1; power4 0; power2 0
}
; Discharge only between 6pm and 4am
if ((min>minrecnct) and (pwr[4]==0) and (time<240 or time>1080)) {
=>backlog power2 1; power4 1; power2 0
}
}

>J
,"Battery":{
"Cell-1":%cv[1]%,
"Cell-2":%cv[2]%,
"Cell-3":%cv[3]%,
"Cell-4":%cv[4]%,
"Cell-5":%cv[5]%,
"Cell-6":%cv[6]%,
"Cell-7":%cv[7]%,
"Cell-8":%cv[8]%,
"Cell-9":%cv[9]%,
"Cell-10":%cv[10]%,
"Cell-11":%cv[11]%,
"Cell-12":%cv[12]%,
"Max":%max%,
"Min":%min%
}

>W
<center>c1=%3cv[1]%, c2=%3cv[2]%<br>
<center>c3=%3cv[3]%, c4=%3cv[4]%<br>
<center>c5=%3cv[5]%, c6=%3cv[6]%<br>
<center>c7=%3cv[7]%, c8=%3cv[8]%<br>
<center>c9=%3cv[9]%, c10=%3cv[10]%<br>
<center>c11=%3cv[11]%, c12=%3cv[12]%<br>
<center>Max=Cell-%maxc% %max%, Min=Cell-%minc% %min%<br>
<center>Avg=%avg%, DCC=%vstr% <br>
<center>DCC=%dcc%, Cntr=%rlcll2%<br>
<center>RlCll3=%3rlcll3%<br>
```
Front page looks like this:
<img width="531" alt="Screen Shot 2022-08-18 at 21 00 37" src="https://user-images.githubusercontent.com/79770016/185493982-4985d464-6f68-4f56-a363-0ab757f2bbd9.png">

Future development:
  - Add a few more arrays to read temperature sensors on the chip
  - Figure out the daisychain option for LTC6802-1 to control more than one battery with a single ESP8266
  - Upgrade wiring and relays to handle 30a
  - A fuses and code to shut off load if higher than 15a
  - Try with ESP32 to have more GPIO available for a seven segment display to show voltage and current
  - Add DS18B20 to measure temps of the MOSFETs and battery cells (would need to swap pin assignments since won't work on GPIO16)
  - Add PWM solar input based on load to offload the battery once it's full. Would require multiple FRN120 in parallel plus some capacitors.

**Comment by barbudor** [comment]
Nice work

**Comment by joba-1** [comment]
Very nice! I love the elegant idea of hybrid relays with mosfets to prevent arcs when switching strong dc. Didn‘t know about this.

Just two notes:
* your [link to the excellent article](https://sound-au.com/articles/hybrid-relays.htm#s2) does not work … yet.
* resist to use that idea for the necessary safety switches. Mosfets tend to fail by creating a short circuit so would defeat the purpose of the safety switch and we are back to your disclaimer. Probably obvious to you but not for all readers :)

**Reply by pkkrusty** [reply]
1. Fixed, thanks
2. Yes, I'm working on notifications to alert me when current is incoming even when solar relays are supposedly off. Putting two MOSFETs in series reduces chance of failure by half... Or maybe just a contactor in between the solar and relays that only opens upon MOSFET failure.

**Comment by norbert-walter** [comment]
Thanks for the explanations about your project. I now have an idea of how you did it. I would like to implement something similar for a boat battery. I have already realized a project with a PZEM-017. https://open-boat-projects.org/en/wifi-battery-monitor/
In the future I would use an INA226. It can measure current in both directions. With the LTC6802-1 I would want to measure the cell voltages. I would build the circuit myself.

**Comment by pkkrusty** [comment]
@gemu2015 Trying this setup with 12.3.1.5 and SPI communication is not working. CS and CLK are performing correctly, and it seems like MOSI is sending correctly, but MISO isn't receiving anything. Downgraded back to 12.0.2.1 with same script and config, and everything works. Haven't stepped up the versions to see where the breaking change is yet, but in the process.

**Reply by pkkrusty** [reply]
OK, it works with 12.2.0, but breaks on 12.3.1.

**Reply by gemu2015** [reply]
i compared the versions according to spi and there is no difference in the spi code.
i connected MOSI with MISO and got everything back as it should.

try narrowing down where the error actually happens, it is not in spi

**Comment by Jason2866** [comment]
Nice! When leaving test status I would replace any breadboard. This will fail sooner or later in real life use.

**Reply by pkkrusty** [reply]
Yeah, I should take the time to solder things together on a perf-board.
